python/ray/serve/_private/router.py

```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
                 )
                 self.in_flight_queries[replica].add(tracker_ref)
             return user_ref
+
+        # Try to find a replica that can handle this query.
+        # If model id is not specified, we can assign it to any non-overladed replica.
+        # If model id is specified, we can try to assign it to a replica that has
+        # loaded, and if there is no such replica, we can assign it to any
+        # non-overloaded replica.
+        if query.metadata.model_id:
+            # Try to find all replicas that has loaded the model and is not overloaded.
+            candidate_replicas = [
```

Comment:
This policy seems to work best if a new model is loaded due to only a single request.

However, suppose a few different end users requests a new model all at once. Ideally, one replica should load that model and serve the few users, just like what happens if they send their requests one after another. Instead, those requests will be sent to different HTTP proxies, and those proxies will assign the requests to many different replicas (since no replica currently has the model loaded). All those replicas may evict their current model when loading the new model. This could cause some thrashing if there’s a couple groups of end users, each requesting a different model.

In other words, if n clients request a new model at roughly the same time, n replicas will evict their current model and load the new one. Ideally, if n is small, only 1 replica should load the model and serve all n clients.

I think this is a relatively common usage pattern, so we should be aware of it. I'm not sure if it's worth handling right now versus in the future.